### ISSUE TITLE
[BOUNTY][help wanted] react-native-webrtc could not process the sdp from gstreamer 1.14.2

### ISSUE THREAD

**Issue by wingchen** [oney/react-native-webrtc#495]
We are trying to connect a stream some video from an embedded system running gstreamer 1.14.2 with mobile running a react-native-webrtc powered project. However, so far we don't have a lot of success with the sdp from gstreamer.

Here comes the sdp from gstreamer:

```
"v=0
o=- 4812078363948229490 0 IN IP4 0.0.0.0
s=-
t=0 0
a=ice-options:trickle
a=msid-semantic:WMS webrtcbin8
m=video 9 UDP/TLS/RTP/SAVPF 97
c=IN IP4 0.0.0.0
a=setup:actpass
a=ice-ufrag:GPT5iqiPEQO4iADBt9+eg3T14oxEg5jJ
a=ice-pwd:xD6u6ewoAWKuuu2InxkTXQng4LPo39Xh
a=sendrecv
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:97 H264/90000
a=rtcp-fb:97 nack pli
a=framerate:24
a=fmtp:97 packetization-mode=1;profile-level-id=420028;sprop-parameter-sets=Z0IAKKlQFAek3AQEBQAAAwPoAAC7gAsIhGoAAA==,aM48MAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==
a=ssrc:3098288572 msid:user2891516485@host-363d171d webrtctransceiver16
a=ssrc:3098288572 cname:user2891516485@host-363d171d
a=mid:video0
a=fingerprint:sha-256 4A:38:BB:55:22:C3:74:24:0F:A5:BA:FC:4F:B8:A2:55:38:50:8F:3C:07:4B:F4:9E:4A:66:0C:2B:9D:82:AB:3D
m=audio 9 UDP/TLS/RTP/SAVPF 96
c=IN IP4 0.0.0.0
a=setup:actpass
a=ice-ufrag:SDMtrdimhGH4DjzH74jLnyK3imbD83YK
a=ice-pwd:k1LIOSghwNVVPry+q6hkpvtzh03FBqV6
a=sendrecv
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 OPUS/48000/2
a=rtcp-fb:96 nack pli
a=fmtp:96 sprop-maxcapturerate=8000;sprop-stereo=0
a=ssrc:2860890805 msid:user2891516485@host-363d171d webrtctransceiver17
a=ssrc:2860890805 cname:user2891516485@host-363d171d
a=mid:audio1
a=fingerprint:sha-256 4A:38:BB:55:22:C3:74:24:0F:A5:BA:FC:4F:B8:A2:55:38:50:8F:3C:07:4B:F4:9E:4A:66:0C:2B:9D:82:AB:3D
"
```

It always returns channel error here. However, we were able to get the demo code working with the sdp from a different mobile device:

```
"v=0
o=- 5502086579047732191 3 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE audio video data
a=msid-semantic: WMS 20F6B101-ED19-4237-98A7-0E5391DB1D1E
m=audio 64072 UDP/TLS/RTP/SAVPF 111 103 104 9 102 0 8 106 105 13 110 112 113 126
c=IN IP4 220.136.74.75
a=rtcp:9 IN IP4 0.0.0.0
a=candidate:178099364 1 udp 2122260223 192.168.55.66 51806 typ host generation 0 network-id 1 network-cost 10
a=candidate:2648053608 1 udp 2122194687 192.168.55.10 64072 typ host generation 0 network-id 2 network-cost 10
a=candidate:480121308 1 udp 1685987071 220.136.74.75 64072 typ srflx raddr 192.168.55.10 rport 64072 generation 0 network-id 2 network-cost 10
a=candidate:2346029584 1 udp 1686052607 220.136.74.75 51806 typ srflx raddr 192.168.55.66 rport 51806 generation 0 network-id 1 network-cost 10
a=candidate:1142843476 1 tcp 1518280447 192.168.55.66 62480 typ host tcptype passive generation 0 network-id 1 network-cost 10
a=candidate:3545623448 1 tcp 1518214911 192.168.55.10 62481 typ host tcptype passive generation 0 network-id 2 network-cost 10
a=ice-ufrag:8VAM
a=ice-pwd:I23xJDu4dRNC4rz7Fo/0+zQ7
a=ice-options:trickle renomination
a=fingerprint:sha-256 F4:E9:0F:74:E3:EB:5D:54:58:57:D2:02:F7:7F:A7:5D:B9:08:6F:48:A0:5D:9C:6C:57:E2:43:AB:AD:3D:33:16
a=setup:actpass
a=mid:audio
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=sendrecv
a=rtcp-mux
a=rtpmap:111 opus/48000/2
a=rtcp-fb:111 transport-cc
a=fmtp:111 minptime=10;useinbandfec=1
a=rtpmap:103 ISAC/16000
a=rtpmap:104 ISAC/32000
a=rtpmap:9 G722/8000
a=rtpmap:102 ILBC/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:106 CN/32000
a=rtpmap:105 CN/16000
a=rtpmap:13 CN/8000
a=rtpmap:110 telephone-event/48000
a=rtpmap:112 telephone-event/32000
a=rtpmap:113 telephone-event/16000
a=rtpmap:126 telephone-event/8000
a=ssrc:3859880680 cname:t7qq+mN7ND6bbLi7
a=ssrc:3859880680 msid:20F6B101-ED19-4237-98A7-0E5391DB1D1E F2230E41-FC25-4B71-A6C8-6AC68E5264F8
a=ssrc:3859880680 mslabel:20F6B101-ED19-4237-98A7-0E5391DB1D1E
a=ssrc:3859880680 label:F2230E41-FC25-4B71-A6C8-6AC68E5264F8
m=video 9 UDP/TLS/RTP/SAVPF 96 97 98 99 100 101 127 125 124
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:8VAM
a=ice-pwd:I23xJDu4dRNC4rz7Fo/0+zQ7
a=ice-options:trickle renomination
a=fingerprint:sha-256 F4:E9:0F:74:E3:EB:5D:54:58:57:D2:02:F7:7F:A7:5D:B9:08:6F:48:A0:5D:9C:6C:57:E2:43:AB:AD:3D:33:16
a=setup:actpass
a=mid:video
a=extmap:2 urn:ietf:params:rtp-hdrext:toffset
a=extmap:3 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:4 urn:3gpp:video-orientation
a=extmap:5 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:6 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=extmap:7 http://www.webrtc.org/experiments/rtp-hdrext/video-content-type
a=extmap:8 http://www.webrtc.org/experiments/rtp-hdrext/video-timing
a=recvonly
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 VP8/90000
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=rtcp-fb:96 goog-remb
a=rtcp-fb:96 transport-cc
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=96
a=rtpmap:98 VP9/90000
a=rtcp-fb:98 ccm fir
a=rtcp-fb:98 nack
a=rtcp-fb:98 nack pli
a=rtcp-fb:98 goog-remb
a=rtcp-fb:98 transport-cc
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=98
a=rtpmap:100 red/90000
a=rtpmap:101 rtx/90000
a=fmtp:101 apt=100
a=rtpmap:127 ulpfec/90000
a=rtpmap:125 H264/90000
a=rtcp-fb:125 ccm fir
a=rtcp-fb:125 nack
a=rtcp-fb:125 nack pli
a=rtcp-fb:125 goog-remb
a=rtcp-fb:125 transport-cc
a=fmtp:125 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f
a=rtpmap:124 rtx/90000
a=fmtp:124 apt=125
m=application 9 DTLS/SCTP 5000
c=IN IP4 0.0.0.0
a=ice-ufrag:8VAM
a=ice-pwd:I23xJDu4dRNC4rz7Fo/0+zQ7
a=ice-options:trickle renomination
a=fingerprint:sha-256 F4:E9:0F:74:E3:EB:5D:54:58:57:D2:02:F7:7F:A7:5D:B9:08:6F:48:A0:5D:9C:6C:57:E2:43:AB:AD:3D:33:16
a=setup:actpass
a=mid:data
a=sctpmap:5000 webrtc-datachannel 1024
"
```

Any clue here?
We are willing to offer BOUNTY for it. Please let me know how you wanna proceed with this (darwing.chen@gmail.com). Thanks a lot

**Comment by ruddell** [comment]
Just a guess, it may be related to the H264 profile level.  From gstreamer I see `420028`, the working SDP uses `42e01f` (baseline).

Chrome 67 offers the following H264 profile levels it its SDP: `42001f`, `42e01f`, `4d0032`, `640032`

**Comment by wingchen** [comment]
@ruddell Thanks for answering. `420028` will potentially be the reason why the frames won't show. However, I don't think it's the current blocker though.

We also tried out building up the native code to see more debugging messages (https://webrtc.org/native-code/ios/). The message we got was:

```
Failed to set remote offer sdp: The BUNDLE group c…data
a=msid-semantic:WMS matching no m= section.
```

It failed at SDP exchange level already.

**Comment by saghul** [comment]
@wingchen Can you enable BUNDLE on that device?

**Comment by wingchen** [comment]
@saghul So far as I know, gstreamer currently does not support bundle. So I think it's a no here.

However, we are going to try this two commits out:
https://github.com/zxcpoiu/react-native-webrtc/commit/c86aea1782d725df752d2c9725151326d549f393/
https://github.com/zxcpoiu/react-native-webrtc/commit/53ffee146864c66d30ce2f7df7b9c348fcd28dad

It will be nice if we are able to get the balance mode explicitly set and get it working. What do you guys think?

I will report here with my findings. Thanks

**Comment by wingchen** [comment]
@saghul @zxcpoiu @ruddell 

Thank you guys. It turns out that the previous commits work:

https://github.com/zxcpoiu/react-native-webrtc/commit/53ffee146864c66d30ce2f7df7b9c348fcd28dad

https://github.com/zxcpoiu/react-native-webrtc/commit/c86aea1782d725df752d2c9725151326d549f393/

We are now forking our own version of react-native-webrtc. But I think it will be nice to have these 2 commits merged in too.

The only reminding issue from us is that we still haven't seen video streaming out from android (but the audio streaming works). I will report here if we have any new update about fixing it.

**Comment by saghul** [comment]
@wingchen Those can be configured, in theory. If we haven't implemented it in the plugin, a PR is definitely welcome!

OTOH, I'd suggest you also push upstream so gstreamer implements BUNDLE and RTCP muxing, in case they don't already. This essentially means using a single port for all audio / video data (both RTC and RTCP), which is a great thing.

**Comment by wingchen** [comment]
@saghul Yep. We are taking to gstreamer community to make it happen. We are also thinking about sponsoring it. But we have to get beyond the end to end production integration first. I like bundle mode better too.

I also got something to report about the android version. It turned out that android version works out time too. The reason the frames were not showing is because we tried to rotate the screen to landscape mode and some of the android machines did not support it. We will change the design on those devices and move on for now.

I am closing this ticket. Thanks
